src/main/java/org/truffleruby/core/numeric/BignumNodes.java

```diff
@@ -700,20 +701,37 @@ public Object compareFallback(DynamicObject a, DynamicObject b) {
     @Primitive(name = "bignum_pow")
     public static abstract class BignumPowPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
-        private final ConditionProfile negativeProfile = ConditionProfile.createBinaryProfile();
+        private static final int BIGLEN_LIMIT = 32 * 1024 * 1024;
+
+        public static BignumPowPrimitiveNode create() {
+            return BignumNodesFactory.BignumPowPrimitiveNodeFactory.create(null);
+        }
+
+        public abstract Object executePow(Object a, Object b);
 
         @Specialization
         public Object pow(DynamicObject a, int b) {
-            return pow(a, (long) b);
+            return executePow(a, (long) b);
         }
 
         @Specialization
-        public Object pow(DynamicObject a, long b) {
+        public Object pow(DynamicObject a, long b,
+                @Cached("createBinaryProfile()") ConditionProfile negativeProfile,
+                @Cached("createBinaryProfile()") ConditionProfile maybeTooBigProfile,
+                @Cached("new()") WarnNode warnNode) {
             if (negativeProfile.profile(b < 0)) {
                 return FAILURE;
             } else {
+                final BigInteger value = Layouts.BIGNUM.getValue(a);
+                final int xbits = value.bitLength();
```

Comment:
Should be `aBits` or `baseBitLength` or so to make it clear it's the bitLength of the base not the exponent.

Reply:
Good call. Our argument name isn't `x`.